Zilon.Core/Zilon.Emulation.Common/AutoplayEngineBase.cs

```diff
@@ -0,0 +1,271 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+
+using Microsoft.Extensions.DependencyInjection;
+
+using Zilon.Bot.Players;
+using Zilon.Bot.Sdk;
+using Zilon.Core.Persons;
+using Zilon.Core.Players;
+using Zilon.Core.Props;
+using Zilon.Core.Schemes;
+using Zilon.Core.Tactics;
+
+namespace Zilon.Emulation.Common
+{
+    public abstract class AutoplayEngineBase<T> where T : IPluggableActorTaskSource
+    {
+        private const int ITERATION_LIMIT = 40_000;
+
+        private bool _changeSector;
+        protected readonly BotSettings _botSettings;
+
+        protected IServiceScope ServiceScope { get; set; }
+
+        public AutoplayEngineBase(BotSettings botSettings)
+        {
+            _botSettings = botSettings;
+        }
+
+        public async Task StartAsync(IServiceProvider _globalServiceProvider)
+        {
+            if (_globalServiceProvider is null)
+            {
+                throw new ArgumentNullException(nameof(_globalServiceProvider));
+            }
+
+            var humanActor = await CreateSectorAsync(_globalServiceProvider).ConfigureAwait(false);
+            var gameLoop = ServiceScope.ServiceProvider.GetRequiredService<IGameLoop>();
+            var botActorTaskSource = _globalServiceProvider.GetRequiredService<T>();
+            botActorTaskSource.Configure(_botSettings);
+
+            var iterationCounter = 1;
+            while (!humanActor.Person.Survival.IsDead && iterationCounter <= ITERATION_LIMIT)
+            {
+                try
+                {
+                    gameLoop.Update();
+
+                    if (_changeSector)
+                    {
+                        humanActor = await CreateSectorAsync(_globalServiceProvider).ConfigureAwait(false);
+
+                        gameLoop = ServiceScope.ServiceProvider.GetRequiredService<IGameLoop>();
+                        botActorTaskSource = ServiceScope.ServiceProvider.GetRequiredService<T>();
+                        botActorTaskSource.Configure(_botSettings);
+
+                        _changeSector = false;
+                    }
+                }
+                catch (ActorTaskExecutionException exception)
+                {
+                    CatchActorTaskExecutionException(exception);
+                }
+#pragma warning disable CA1031 // Do not catch general exception types
+                catch(Exception exception)
+#pragma warning restore CA1031 // Do not catch general exception types
+                {
+                    CatchException(exception);
+                }
+            };
+
+            ProcessEnd();
+        }
+
+        protected abstract void CatchException(Exception exception);
+
+        protected abstract void CatchActorTaskExecutionException(ActorTaskExecutionException exception);
+
+        protected abstract void ProcessEnd();
+
+        private static IActor CreateHumanActor(HumanPlayer humanPlayer,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Change the visibility of this constructor to 'protected'.](https://app.codacy.com/manual/kreghek/Zilon_Roguelike/pullRequest?prid=5028977)